File: examples/python/01_list_tables.py
"""
Example 1: Connect + List Tables
Fully standalone — copy this file and run it.
"""

import struct
import pandas as pd
import pyodbc
from azure.identity import AzureCliCredential, ClientSecretCredential
from azure.keyvault.secrets import SecretClient

def connect_to_fabric():
    vault_url = "https://uzima-secrets-xfmh.vault.azure.net"
    kv_tenant = "4fde8ff3-4dd5-42e1-a25a-e42905610d66"
    server    = "fis5jjpzajqe5fxqs4z3vlsjde-zgopmz6jacoezkc3hd6da52lpm.datawarehouse.fabric.microsoft.com"

    kv_client = SecretClient(vault_url, AzureCliCredential(tenant_id=kv_tenant))
    tenant_id = kv_client.get_secret("fabric-sp-tenant-id").value
    client_id = kv_client.get_secret("fabric-sp-client-id").value
    client_secret = kv_client.get_secret("fabric-sp-client-secret").value

    token = ClientSecretCredential(tenant_id, client_id, client_secret
        ).get_token("https://database.windows.net/.default").token

    token_bytes = token.encode("utf-16-le")
    token_struct = struct.pack(f"<I{len(token_bytes)}s", len(token_bytes), token_bytes)

    conn_str = (
        "Driver={ODBC Driver 18 for SQL Server};"
        f"Server=tcp:{server},1433;"
        "Database=uzima_db_backup;"
        "Encrypt=yes;TrustServerCertificate=no;"
    )
    return pyodbc.connect(conn_str, attrs_before={1256: token_struct}, timeout=30)


conn = connect_to_fabric()
print("Connected!\n")

tables = pd.read_sql_query(
    "SELECT TABLE_SCHEMA, TABLE_NAME FROM INFORMATION_SCHEMA.TABLES ORDER BY TABLE_SCHEMA, TABLE_NAME",
    conn
)
print(tables.to_string(index=False))
conn.close()

File: examples/R/01_list_tables.R
# ============================================================================
# Example 1: Connect + List Tables
#
# Auth flow: Device code (browser) → Key Vault → SP → ODBC SQL
#
# Prerequisites:
#   install.packages(c("httr", "jsonlite", "odbc", "DBI"))
#   ODBC Driver 18 for SQL Server
#   remotes::install_github("AKU-CDIO/fabric-inbound-access", subdir = "fabriconnect")
# ============================================================================

rm(list = ls())

library(httr)
library(jsonlite)
library(odbc)
library(DBI)
library(fabriconnect)

# ---- Config ----
VAULT_URL    <- "https://uzima-secrets-xfmh.vault.azure.net"
SERVER       <- "fis5jjpzajqe5fxqs4z3vlsjde-zgopmz6jacoezkc3hd6da52lpm.datawarehouse.fabric.microsoft.com"
DATABASE     <- "uzima_db_backup"

# ---- Step 1: Device code → browser login → get token ----
cat("Signing in via browser...\n")
cfg <- jsonlite::fromJSON(system.file("config.json", package = "fabriconnect"))
token <- fabriconnect:::.get_fabric_token(cfg$fabric_tenant)
cat("Authenticated.\n\n")

# ---- Step 2: Token → Key Vault → get SP credentials ----
cat("Fetching SP credentials from Key Vault...\n")
fetch_secret <- function(name) {
  url <- paste0(VAULT_URL, "/secrets/", name, "?api-version=7.4")
  resp <- GET(url, add_headers(Authorization = paste("Bearer", token)))
  stop_for_status(resp)
  content(resp)$value
}

tenant_id     <- fetch_secret("fabric-sp-tenant-id")
client_id     <- fetch_secret("fabric-sp-client-id")
client_secret <- fetch_secret("fabric-sp-client-secret")
cat("SP credentials retrieved.\n\n")

# ---- Step 3: SP credentials → SQL token ----
cat("Getting SQL access token...\n")
resp <- POST(
  paste0("https://login.microsoftonline.com/", tenant_id, "/oauth2/v2.0/token"),
  body = list(
    grant_type    = "client_credentials",
    client_id     = client_id,
    client_secret = client_secret,
    scope         = "https://database.windows.net/.default"
  ),
  encode = "form"
)
stop_for_status(resp)
sql_token <- content(resp)$access_token
cat("SQL token obtained.\n\n")

# ---- Step 4: SQL token → ODBC connection ----
cat("Connecting to Fabric SQL...\n")
con <- dbConnect(
  odbc::odbc(),
  Driver      = "ODBC Driver 18 for SQL Server",
  Server      = paste0(SERVER, ",1433"),
  Database    = DATABASE,
  UID         = 1,
  AccessToken = sql_token,
  Encrypt     = "yes",
  TrustServerCertificate = "no",
  Timeout     = 30
)
cat("Connected!\n\n")

# ---- List tables ----
tables <- dbGetQuery(con, "SELECT TABLE_SCHEMA, TABLE_NAME FROM INFORMATION_SCHEMA.TABLES ORDER BY TABLE_NAME")
cat("Tables found:", nrow(tables), "\n")
print(tables)

dbDisconnect(con)
